cpp/include/cuml/experimental/fil/decision_forest.hpp

```diff
@@ -0,0 +1,466 @@
+   * @param[in] input The buffer containing the input data
+   * @param[in] stream For GPU execution, the CUDA stream. For CPU execution,
+   * this optional parameter can be safely omitted.
+   * @param[in] specified_rows_per_block_iter If non-nullopt, this value is
+   * used to determine how many rows are evaluated for each inference
+   * iteration within a CUDA block. Runtime performance is quite sensitive
+   * to this value, but it is difficult to predict a priori, so it is
+   * recommended to perform a search over possible values with realistic
+   * batch sizes in order to determine the optimal value. Any power of 2 from
+   * 1 to 32 is a valid value, and in general larger batches benefit from
+   * larger values.
+   */
+  void predict(
+    raft_proto::buffer<typename forest_type::io_type>& output,
+    raft_proto::buffer<typename forest_type::io_type> const& input,
+    raft_proto::cuda_stream stream = raft_proto::cuda_stream{},
+    std::optional<index_type> specified_rows_per_block_iter=std::nullopt
+  ) {
+    if (output.memory_type() != memory_type() || input.memory_type() != memory_type()) {
+      throw raft_proto::wrong_device_type{
+        "Tried to use host I/O data with model on device or vice versa"
+      };
+    }
+    if (output.device_index() != device_index() || input.device_index() != device_index()) {
+      throw raft_proto::wrong_device{
+        "I/O data on different device than model"
+      };
+    }
+    auto* vector_output_data = (
+      vector_output_.has_value() ? vector_output_->data() : static_cast<io_type*>(nullptr)
+    );
+    auto* categorical_storage_data = (
+      categorical_storage_.has_value() ? categorical_storage_->data() : static_cast<categorical_storage_type*>(nullptr)
+    );
+    switch(nodes_.device().index()) {
+      case 0:
+        fil::detail::infer(
+          obj(),
+          get_postprocessor(),
+          output.data(),
+          input.data(),
+          index_type(input.size() / num_features_),
+          num_features_,
+          num_outputs_,
+          has_categorical_nodes_,
+          vector_output_data,
+          categorical_storage_data,
+          specified_rows_per_block_iter,
+          std::get<0>(nodes_.device()),
+          stream
+        );
+        break;
+      case 1:
+        fil::detail::infer(
+          obj(),
+          get_postprocessor(),
+          output.data(),
+          input.data(),
+          index_type(input.size() / num_features_),
+          num_features_,
+          num_outputs_,
+          has_categorical_nodes_,
+          vector_output_data,
+          categorical_storage_data,
+          specified_rows_per_block_iter,
+          std::get<1>(nodes_.device()),
+          stream
+        );
+        break;
+    }
+  }
+
+ private:
+  /** The nodes for all trees in the forest */
+  raft_proto::buffer<node_type> nodes_;
+  /** The index of the root node for each tree in the forest */
+  raft_proto::buffer<index_type> root_node_indexes_;
```

Comment:
I like how you collected all nodes in a single array. (This is different from Treelite, where a forest has a vector of Tree objects, each of which carries a separate node array.)